### PR TITLE
[FLINK-25351] Introduce `FlinkVersion` as a global enum

### PR DESCRIPTION
**Pull request by matriv**
## What is the purpose of the change

Rename `MigrationVersion` to `FlinkVersion` and move it to
`flink-annoations` `org.apache.flink` package so that is available
globally and will facilitate the APIs and SQL/TableAPI versioning
and upgrades.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper:**no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **update [release guide](https://cwiki.apache.org/confluence/display/FLINK/Creating+a+Flink+Release)**

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java [hunk 1]
```diff
@@ -16,7 +16,7 @@
  * limitations under the License.
  */
 
-package org.apache.flink.testutils.migration;
+package org.apache.flink.core;
```
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java:19]: For FLIP-197 this enum must be accessible for `@PublicEvolving` which is in the `flink-annotation` module. Unless we let `flink-annotation` depend on `flink-core`. @tillrohrmann what is your preference here?

We could move the class to the module but does the package `org.apache.flink.annotation` fit? Alternatively, we could also move this to a package `org.apache.flink.FlinkVersion` in the `flink-annotation` module.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks @matriv. I left an important comment.

File: flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java [hunk 2]
```diff
@@ -59,22 +59,22 @@ public String toString() {
...
 
-    public boolean isNewerVersionThan(MigrationVersion otherVersion) {
+    public boolean isNewerVersionThan(FlinkVersion otherVersion) {
         return this.ordinal() > otherVersion.ordinal();
     }
 
     /** Returns all versions equal to or higher than the selected version. */
-    public List<MigrationVersion> orHigher() {
-        return Stream.of(MigrationVersion.values())
+    public List<FlinkVersion> orHigher() {
```
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java:67]: a `Set` might be a better return type here

File: flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java:19]: But I can also move it over once I get to implementing FLIP-197.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @matriv. The change looks good to me modulo what Timo mentioned. We could already move it to `flink-annotations` or we'll tackle it once we start with FLIP-197.

**Comment by matriv** [comment]
@tillrohrmann @twalthr Thank you! Moved to `flink-annotations` please check again.

File: flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java (hunk 1 above)
Comment (matriv) [reply flink-core/src/main/java/org/apache/flink/core/FlinkVersion.java]: @tillrohrmann No problem, please, I've already moved to `flink-annotations`.

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**